File: node/src/scoring.ts
import type { Normalize } from "./config.js";

const PUNCTUATION_RE = /[!"#$%&'()*+,\-./:;<=>?@[\]^_`{|}~]/g;
const WHITESPACE_RE = /\s+/g;

export function normalizeText(text: string, normalize: Normalize): string {
  let result = text.trim();
  if (normalize.caseInsensitive) {
    result = result.toLowerCase();
  }
  if (normalize.stripPunctuation) {
    result = result.replace(PUNCTUATION_RE, "");
  }
  if (normalize.collapseWhitespace) {
    result = result.replace(WHITESPACE_RE, " ").trim();
  }
  return result;
}

export function extractAnswer(text: string, pattern: string | null): string {
  if (!pattern) {
    return text;
  }
  const match = new RegExp(pattern, "s").exec(text);
  if (!match) {
    return text;
  }
  return match.length > 1 ? match[1] : match[0];
}

export interface ScoreResult {
  consistencyScore: number;
  passed: boolean;
  distinctAnswers: number;
  majorityAnswer: string;
  distribution: Record<string, number>;
}

export function scoreResponses(
  rawResponses: string[],
  normalize: Normalize,
  threshold: number,
  extractPattern: string | null = null,
): ScoreResult {
  if (rawResponses.length === 0) {
    throw new Error("cannot score an empty list of responses");
  }

  const normalized = rawResponses.map((r) => normalizeText(extractAnswer(r, extractPattern), normalize));
  const counts = new Map<string, number>();
  for (const n of normalized) {
    counts.set(n, (counts.get(n) ?? 0) + 1);
  }

  let majorityAnswer = normalized[0];
  let majorityCount = 0;
  for (const [answer, count] of counts) {
    if (count > majorityCount) {
      majorityAnswer = answer;
      majorityCount = count;
    }
  }

  const consistencyScore = majorityCount / normalized.length;
  return {
    consistencyScore,
    passed: consistencyScore >= threshold,
    distinctAnswers: counts.size,
    majorityAnswer,
    distribution: Object.fromEntries(counts),
  };
}

File: python/src/flakegate/scoring.py
from __future__ import annotations

import re
import string
from collections import Counter
from dataclasses import dataclass

from flakegate.config import Normalize

_WHITESPACE_RE = re.compile(r"\s+")
_PUNCTUATION_TABLE = str.maketrans("", "", string.punctuation)


def normalize_text(text: str, normalize: Normalize) -> str:
    result = text.strip()
    if normalize.case_insensitive:
        result = result.lower()
    if normalize.strip_punctuation:
        result = result.translate(_PUNCTUATION_TABLE)
    if normalize.collapse_whitespace:
        result = _WHITESPACE_RE.sub(" ", result).strip()
    return result


def extract_answer(text: str, pattern: str | None) -> str:
    if not pattern:
        return text
    match = re.search(pattern, text, re.DOTALL)
    if not match:
        return text
    return match.group(1) if match.groups() else match.group(0)


@dataclass(frozen=True)
class ScoreResult:
    consistency_score: float
    passed: bool
    distinct_answers: int
    majority_answer: str
    distribution: dict[str, int]


def score_responses(
    raw_responses: list[str],
    normalize: Normalize,
    threshold: float,
    extract_pattern: str | None = None,
) -> ScoreResult:
    if not raw_responses:
        raise ValueError("cannot score an empty list of responses")

    normalized = [
        normalize_text(extract_answer(r, extract_pattern), normalize) for r in raw_responses
    ]
    counts = Counter(normalized)
    majority_answer, majority_count = counts.most_common(1)[0]
    consistency_score = majority_count / len(normalized)

    return ScoreResult(
        consistency_score=consistency_score,
        passed=consistency_score >= threshold,
        distinct_answers=len(counts),
        majority_answer=majority_answer,
        distribution=dict(counts),
    )
